Avoid overflow in p norms

andreasnoack opened the pull request:
Use division instead of multiplication with reciprocal when scaling elements in vecnorm methods. This fixes JuliaLang/LinearAlgebra.jl#222 and supersedes JuliaLang/julia#11789. The logic used here is a little different from the other `generic_vecnorm`s. I try to use the element type of the array when the input is an `AbstractVector{T<:Number}` and for general iterables I have to require that they are non-empty.

cc: @stevengj

stevengj commented:
Didn't the old implementation work for `Array{T}` for any `T` that defines `norm(x::T)`?  See #11043.  (I should have added a test case for that, grrr.)  The new implementation loses this because it requires `zero(T)`; I don't think that should be necessary.

andreasnoack commented:
The new version doesn't require `zero(T)`. It only uses `zero(T)` when `T<:Number`. For all other element types it uses `zero(x[1])`.

andreasnoack commented:
...but I'm fine with using `norm` instead of `abs`. I'll change that now.

stevengj commented:
Can you add a test case for the vecnorm of an array of arrays, if we don't have one yet?

stevengj commented on base/linalg/generic.jl:
```diff
@@ -103,21 +103,33 @@ function generic_vecnorm1(x)
+        throw(ArgumentError("input must either be an AbstractVector{T<:Number} or non-empty"))
     end
-    return convert(T, maxabs * sqrt(sum))
+    scale = zero(norm(x[1]))
```
It's kind of annoying to compute `norm(x[1])` twice in this case.  Maybe add an `i0` argument to `generic_vecnorm2` and loop from `i = i0:n`.  That way you can pass `i0 == 2` to avoid recomputing the first element, which can be passed via `ssq`.

andreasnoack replied:
I don't think `norm` will actually be computed here. E.g. for a `Vector{Complex{Float64}}` I get that
```
julia> @code_llvm zero(norm(x2[1]))

define double @julia_zero_21182(double) {
top:
  ret double 0.000000e+00
}
```

yuyichao replied:
`@code_llvm` compute the argument before getting its type.

yuyichao replied:
Even with inbounds,

```julia
julia> a = Complex128[1, 2, 3]
3-element Array{Complex{Float64},1}:                                              
 1.0+0.0im                                                                        
 2.0+0.0im                                                                        
 3.0+0.0im                                                                        

julia> f(x) = (@inbounds v = zero(norm(x[1])); v)
f (generic function with 1 method)

julia> @code_llvm f(a)

define double @julia_f_21134(%jl_value_t*) {
top:
  %1 = alloca %Complex.12, align 8
  %2 = bitcast %jl_value_t* %0 to %Complex.12**
  %3 = load %Complex.12** %2, align 8
  %4 = load %Complex.12* %3, align 8
  store %Complex.12 %4, %Complex.12* %1, align 8
  %5 = call double @julia_norm_21121(%Complex.12* %1)
  ret double 0.000000e+00
}
```

I don't think type inference or llvm is smart enough to figure out `norm(::Complex)` is effect free. There's even more junk emitted without `@inbounds`.

stevengj replied:
I'm mostly worried about it for something like an array of matrices, where `norm` is expensive.

stevengj commented on base/linalg/generic.jl:
```diff
@@ -103,21 +103,33 @@ function generic_vecnorm1(x)
+end
+function generic_vecnorm2{T<:Number}(x::AbstractVector{T})
+    scale = zero(norm(zero(T)))
+    ssq = one(sqrt((abs2(norm(zero(T))) + abs2(norm(zero(T))))/2))
```
Is all this really needed just to set `ssq = 0` of the appropriate type?   Unless the compiler is clever enough to discard all of the extraneous calculations?

StefanKarpinski replied:
It almost is for `Float64`:

```jl
julia> f{T}(x::T) = one(sqrt((abs2(norm(zero(T))) + abs2(norm(zero(T))))/2))
f (generic function with 2 methods)

julia> f(123.45)
1.0

julia> @code_llvm f(123.45)

define double @julia_f_21321(double) {
top:
  %1 = call double @julia_norm_21171(double 0.000000e+00)
  %2 = call double @julia_norm_21171(double 0.000000e+00)
  ret double 1.000000e+00
}
```

So it figures out that the result is constant but still calls the Float64 norm function twice in case there are side effects.

StefanKarpinski commented:
@stevengj, should we just merge this and fix the issues in a separate change?

andreasnoack commented:
Hold on a bit. I'll add the tests @stevengj asked for and there might also be few other adjustments based on the Slack discussion this evening.

staticfloat commented:
@andreasnoack bumping this since it's slated for 0.4.

andreasnoack commented:
I'm on it. I've been reading up on why BLAS implementation and some critique of it. Scaling all elements even when it's not necessary is expensive. I might just fix the overflow problem problem for now and then we can reconsider the algorithm in 0.5.

andreasnoack committed:
User division instead of multiplication with reciprocal in generic_vecnorm2
to avoid potential overflow and fewer rounding errors. Fixes #11789.

Make sure that norm(Number) is inlined.

andreasnoack commented:
This one works now. I've patched the existing versions to handle underflow without being dead slow because of the division, i.e. the value are only scaled when necessary. This is also the reason to avoid the BLAS implementation because it scales even when it is not necessary and therefore it is very slow.

cc: @stevengj

stevengj commented:
LGTM.

stevengj commented:
Oh, darn it, I forgot to check whether you added a test for the array-of-arrays case (you didn't).

It looks like you broke `norm` for this case, because you call `one`, which isn't defined for a general Banach space.  e.g. `norm(Vector{Int}[[1,2], [3,4]])` is broken.

andreasnoack commented:
I'll look into it.

Den torsdag den 3. september 2015 skrev Steven G. Johnson <
notifications@github.com>:

> Oh, darn it, I forgot to check whether you added a test for the
> array-of-arrays case (you didn't).
>
> It looks like you broke norm for this case, because you call one, which
> isn't defined for a general Banach space. e.g. norm(Vector{Int}[[1,2],
> [3,4]]) is broken.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/JuliaLang/julia/pull/12571#issuecomment-137570104>.
>

stevengj commented:
I have PR ready.
